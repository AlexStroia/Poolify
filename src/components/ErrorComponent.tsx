import React from "react";
import { Alert, AlertTitle } from "@mui/material";

const ErrorComponent = (message: string) => {
  return (
    <div
      style={{
        width: "100%",
        maxWidth: "400",
        margin: "0 auto",
        marginTop: "4",
      }}
    >
      <Alert severity="error" variant="outlined">
        <AlertTitle
          style={{
            fontWeight: "bold",
            marginBottom: "1",
          }}
        >
          Error
        </AlertTitle>
        {message}
      </Alert>
    </div>
  );
};

export default ErrorComponent;
