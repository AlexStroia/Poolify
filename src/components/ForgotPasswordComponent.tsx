import { Grid } from "@mui/material";
import { PoolifyTextField } from "../views/PoolifyTextField";
import { PoolifyButton } from "../views/PoolifyButton";
import { PoolifyAppBar } from "../views/PoolifyAppBar";
import { SpinnerComponent } from "./SpinnerComponent";
import { useEffect, useRef, useState } from "react";
import { useSelector } from "react-redux";
import { ApplicationState } from "../state/ApplicationState";
import PoolifySnackbar from "../views/PoolifySnackBar";

const emailSuccessMessage =
  "If that email exist in our database, you should receive an email shortly.";

export const ForgotPasswordComponent = ({
  onTapForgotPassword = (email: string) => {},
  onClose = () => {},
}) => {
  const [showSnackbar, setShowSnackbar] = useState<boolean>(false);
  const email = useRef<HTMLInputElement>(null);

  const { errorMessage: error, loading } = useSelector(
    (state: ApplicationState) => state.authentication,
  );

  const handleTapForgotPassword = () => {
    const emailValue = email.current?.value ?? "";
    onTapForgotPassword(emailValue);
    handleOpenSnackbar();
  };

  const handleOpenSnackbar = () => {
    setShowSnackbar(true);
  };

  const handleCloseSnackbar = () => {
    setShowSnackbar(false);
    onClose();
  };

  return loading ? (
    <SpinnerComponent />
  ) : (
    <Grid
      container
      direction="column"
      justifyContent="center"
      display="flex"
      alignItems="center"
      style={{ height: "100vh", gap: "16px" }}
    >
      <PoolifySnackbar
        open={showSnackbar}
        severity={error ? "error" : "success"}
        message={error ?? emailSuccessMessage}
        onClose={handleCloseSnackbar}
      />
      <PoolifyTextField
        inputRef={email}
        label="Email"
        placeholder="johndoe@gmail.com"
      />
      <PoolifyButton title="Send Password" onTap={handleTapForgotPassword} />
    </Grid>
  );
};
