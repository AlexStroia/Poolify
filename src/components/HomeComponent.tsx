import { Box, Container, Divider, Grid, Typography } from "@mui/material";
import { PoolifyTabBar } from "../views/PoolifyTabBar";
import QuestionList, { QuestionListType } from "../views/QuestionList";
import React, { useEffect, useState } from "react";
import { getAllQuestions } from "../actions/GetAllQuestions";
import { useDispatch, useSelector } from "react-redux";
import { ApplicationState } from "../state/ApplicationState";
import { useNavigate } from "react-router-dom";
import { SpinnerComponent } from "./SpinnerComponent";
import { getUserQuestions } from "../actions/GetUserQuestions";

export const HomeComponent = () => {
  const dispatch = useDispatch();
  const navigator = useNavigate();
  const state = useSelector((state: ApplicationState) => state);
  const dashboardState = state.dashboard;
  const user = state.authentication.user;

  useEffect(() => {
    dispatch(getAllQuestions());
    if (user?.userId !== null) {
      dispatch(getUserQuestions(user!.userId!));
    }
  }, []);

  const handleOnTapQuestion = (questionId: string) => {
    navigator(`/questions/${questionId}`);
  };

  const questionsNewOrdered =
    dashboardState.userNewQuestions !== undefined &&
    Array.isArray(dashboardState.userNewQuestions) &&
    dashboardState.userNewQuestions.length > 0
      ? dashboardState.userNewQuestions
          .slice()
          .sort(
            (first, second) =>
              new Date(second.date).getTime() - new Date(first.date).getTime()
          )
      : [];

  const questionAnsweredOrdered =
    dashboardState.userAnsweredQuestions !== undefined &&
    Array.isArray(dashboardState.userAnsweredQuestions) &&
    dashboardState.userAnsweredQuestions.length > 0
      ? dashboardState.userAnsweredQuestions
          .slice()
          .sort(
            (first, second) =>
              new Date(second.date).getTime() - new Date(first.date).getTime()
          )
      : [];

  return (
    <div>
      <PoolifyTabBar />
      <Container
        sx={{
          flexDirection: "column",
          gap: "124px", // Increase the spacing between elements to 24px
        }}
      >
        <Box height="32px" />
        {dashboardState.loadingNewQuestions ? (
          <SpinnerComponent />
        ) : (
          <Grid
          >
            <Grid 
                      data-testid="questions-list-new"

            item>
              <QuestionList
                questionListType={QuestionListType.NEW}
                questionDataList={questionsNewOrdered}
                onTapQuestion={handleOnTapQuestion}
              />
            </Grid>

            <Grid 
                      data-testid="questions-list-done"
            item>
              <QuestionList
                questionListType={QuestionListType.DONE}
                questionDataList={questionAnsweredOrdered}
                onTapQuestion={handleOnTapQuestion}
              />
            </Grid>
          </Grid>
        )}
      </Container>
    </div>
  );
};