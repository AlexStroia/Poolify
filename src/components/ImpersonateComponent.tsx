import { Typography, Grid, Card } from "@mui/material";
import { PoolifyButton } from "../views/PoolifyButton";
import React from "react";

export const ImpersonateComponent = ({
  onTapUdacityAccount,
  onTapAlexAccount,
}: {
  onTapUdacityAccount: () => void;
  onTapAlexAccount: () => void;
}) => {
  const handleOnTapUdacityAccount = () => {
    onTapUdacityAccount();
  };

  const handleOnTapAlexAccount = () => {
    onTapAlexAccount();
  };

  return (
    <Card
      sx={{
        padding: "16px",
      }}
    >
      <Grid
        container
        direction="column"
        justifyContent="center"
        alignItems="center"
        alignContent="center"
        style={{
          gap: "16px",
        }}
      >
        <Grid item>
          <Typography>Choose an account from below</Typography>
        </Grid>
        <Grid
          sx={{
            justifyContent: "center",
            display: "flex",
          }}
          item
        >
          <PoolifyButton title="Udacity" onTap={handleOnTapUdacityAccount} />
        </Grid>
        <Grid
          sx={{
            justifyContent: "center",
            display: "flex",
          }}
          item
          onClick={onTapAlexAccount}
        >
          <PoolifyButton title="Alex" onTap={handleOnTapAlexAccount} />
        </Grid>
      </Grid>
    </Card>
  );
};

export default ImpersonateComponent;
