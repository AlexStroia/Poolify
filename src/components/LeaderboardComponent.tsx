import { PoolifyTabBar } from "../views/PoolifyTabBar";

export const LeaderboardComponent = () => {
  return <PoolifyTabBar />;
};
