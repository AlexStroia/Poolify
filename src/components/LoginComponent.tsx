import { Grid, Typography } from "@mui/material";
import {
  PoolifyTextField,
  PoolifyTextFieldInputType,
} from "../views/PoolifyTextField";
import { PoolifyButton } from "../views/PoolifyButton";
import { TransparentButton } from "../views/TransparentButton";
import { PoolifyAppBar } from "../views/PoolifyAppBar";
import { theme } from "../theme";
import { useRef } from "react";

export const LoginComponent = ({
  onTapSignup = () => {},
  onTapForgotPassword = () => {},
}) => {
  const email = useRef(null);
  const password = useRef(null);
  return (
    <Grid
      container
      direction="column"
      justifyContent="center"
      display="flex"
      alignItems="center"
      style={{ height: "100vh" }}
    >
      <PoolifyTextField
        label="Email"
        placeholder="johndoe@gmail.com"
        inputRef={email}
      />
      <PoolifyTextField
        label="Password"
        placeholder="123456"
        inputType={PoolifyTextFieldInputType.Password}
        inputRef={password}
      />
      <PoolifyButton
        title="Login"
        onTap={function (): {} {
          throw new Error("Function not implemented.");
        }}
      />
      <Typography>
        Do not have an account?{" "}
        <TransparentButton title={"Sign Up"} onTap={onTapSignup} />
      </Typography>

      <Typography
        onClick={onTapForgotPassword}
        sx={{
          color: theme.palette.primary.main,
        }}
      >
        Forgot Password
      </Typography>
    </Grid>
  );
};
