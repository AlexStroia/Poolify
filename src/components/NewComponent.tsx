import { Grid, Typography, Box, Card } from "@mui/material";
import { PoolifyTextField } from "../views/PoolifyTextField";
import { PoolifyButton } from "../views/PoolifyButton";
import { useRef } from "react";
import { useDispatch, useSelector } from "react-redux";
import {
  SaveUserQuestionData,
  saveUserQuestion,
} from "../actions/SaveUserQuestion";
import { ApplicationState } from "../state/ApplicationState";
import { SpinnerComponent } from "./SpinnerComponent";
import { PoolifyTabBar } from "../views/PoolifyTabBar";

export const NewComponent = () => {
  const dispatch = useDispatch();
  const questionOptionFirst = useRef<HTMLInputElement>(null);
  const questionOptionSecond = useRef<HTMLInputElement>(null);
  const user = useSelector(
    (state: ApplicationState) => state.authentication.user,
  );
  const dashboardState = useSelector(
    (state: ApplicationState) => state.dashboard,
  );

  const handleTapSave = () => {
    const questionOptionFirstValue = questionOptionFirst.current?.value ?? "";
    const questionOptionSecondValue = questionOptionSecond.current?.value ?? "";
    const saveUserQuestionData: SaveUserQuestionData = {
      questionOptionFirst: questionOptionFirstValue,
      questionOptionSecond: questionOptionSecondValue,
      date: new Date().toISOString(),
      userId: user?.userId ?? "",
      email: user?.email ?? "",
    };
    dispatch(
      saveUserQuestion({
        userId: user?.userId,
        saveUserQuestionData: saveUserQuestionData,
      }),
    );
  };

  return dashboardState.loading ? (
    <SpinnerComponent />
  ) : (
    <div>
      <PoolifyTabBar />
      <Card
        style={{ marginTop: "16px", borderRadius: "16px", padding: "16px" }}
      >
        <div>
          <Grid container direction="column" spacing={2} display="flex">
            {" "}
            <Grid
              item
              sx={{
                justifyContent: "center",
                display: "flex",
              }}
            >
              {" "}
              <Typography variant="h4" gutterBottom>
                Would you rather?
              </Typography>
            </Grid>
            <Grid
              item
              sx={{
                justifyContent: "center",
                display: "flex",
              }}
            >
              <PoolifyTextField
                label="First"
                placeholder="3"
                inputRef={questionOptionFirst}
              />
            </Grid>
            <Grid
              item
              sx={{
                justifyContent: "center",
                display: "flex",
              }}
            >
              <PoolifyTextField
                label="Second"
                placeholder="3"
                inputRef={questionOptionSecond}
              />
            </Grid>
            <Grid
              item
              sx={{
                justifyContent: "center",
                display: "flex",
              }}
            >
              <PoolifyButton title="Save" onTap={handleTapSave} />
            </Grid>
          </Grid>
        </div>
      </Card>
    </div>
  );
};

export default NewComponent;
