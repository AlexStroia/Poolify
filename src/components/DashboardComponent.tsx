import { Box, Grid, Tab, Tabs, Typography } from "@mui/material";
import { theme } from "../theme";
import { TabPanelContent } from "../views/TabPanelContent";
import { HomeComponent } from "./HomeComponent";
import { LeaderboardComponent } from "./LeaderboardComponent";
import { NewComponent } from "./NewComponent";
import { PoolifyButton } from "../views/PoolifyButton";
import { useDispatch, useSelector } from "react-redux";
import { ApplicationState } from "../state/ApplicationState";
import { logoutAction } from "../actions/LogoutAction";
import { changePage } from "../reducers/DashboardSlice";
import { DashboardPage } from "../state/DashboardState";

export const DashboardComponent = () => {
  const dispatch = useDispatch();
  const state = useSelector((state: ApplicationState) => state);

  const user = state.authentication.user;
  const page = state.dashboard.page;

  const handleTabChange = (page: DashboardPage) => {
    dispatch(changePage(page));
  };

  const handleLogout = () => {
    dispatch(logoutAction());
  };

  return (
    <div>
      <Grid container alignItems="center">
        <Grid item sx={{ flex: 1 }}>
          <Box sx={{ borderColor: theme.palette.primary.main }}>
            <Tabs
              value={page}
              onChange={(event, newPage) => {
                handleTabChange(newPage);
              }}
            >
              <Tab label="Home" />
              <Tab label="Leaderboard" />
              <Tab label="New" />
            </Tabs>
          </Box>
        </Grid>
        <Grid item>
          <Typography variant="body2" sx={{ marginRight: 2 }}>
            {user?.email}
          </Typography>
        </Grid>
        <Grid item>
          <PoolifyButton title="Logout" onTap={handleLogout} />
        </Grid>
      </Grid>
      <TabPanelContent value={page} index={0} component={<HomeComponent />} />
      <TabPanelContent
        value={page}
        index={1}
        component={<LeaderboardComponent />}
      />
      <TabPanelContent value={page} index={2} component={<NewComponent />} />
    </div>
  );
};
