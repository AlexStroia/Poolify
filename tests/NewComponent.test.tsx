import React from "react";
import { render } from "@testing-library/react";
import { NewComponent } from "../src/components/NewComponent";
import { BrowserRouter } from "react-router-dom";

jest.mock("react-redux", () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn().mockReturnValue({
    authentication: { user: { userId: "test_user_id", email: "test_user_email" } },
    dashboard: {
      loading: false,
      errorMessage: null,
    },
  }),
}));

describe("NewComponent", () => {
  it("matches the snapshot", () => {
    const { asFragment } = render(
      <BrowserRouter> {/* Wrap NewComponent with BrowserRouter */}
        <NewComponent />
      </BrowserRouter>
    );    expect(asFragment()).toMatchSnapshot();
  });
});
