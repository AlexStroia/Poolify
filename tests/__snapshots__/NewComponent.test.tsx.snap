// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewComponent matches the snapshot 1`] = `
<DocumentFragment>
   
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-1vpwcmr-MuiGrid-root"
    style="height: 100vh; width: 100%;"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container css-1vam7s3-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item css-1s76rlx-MuiGrid-root"
        >
          <div
            class="MuiBox-root css-1h5pmnd"
          >
            <div
              class="MuiTabs-root css-1ujnqem-MuiTabs-root"
            >
              <div
                class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
                style="overflow: hidden; margin-bottom: 0px;"
              >
                <div
                  class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
                  role="tablist"
                >
                  <a
                    aria-selected="false"
                    class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                    href="/home"
                    role="tab"
                    tabindex="-1"
                  >
                    Home
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </a>
                  <a
                    aria-selected="false"
                    class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                    href="/leaderboard"
                    role="tab"
                    tabindex="-1"
                  >
                    Leaderboard
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </a>
                  <a
                    aria-selected="false"
                    class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                    href="/add"
                    role="tab"
                    tabindex="-1"
                  >
                    New
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </a>
                </div>
                <span
                  class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
                  style="left: 0px; width: 0px;"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body2 css-1g8k7mg-MuiTypography-root"
          />
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-u4gkmc-MuiButtonBase-root-MuiButton-root"
            data-testid="show-button"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiTypography-root MuiTypography-button css-1bvnr3d-MuiTypography-root"
            >
              Logout
            </span>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-z7x17r-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
        style="margin-top: 16px; border-radius: 16px; padding: 16px;"
      >
        <div>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column css-ymt1gn-MuiGrid-root"
          >
             
            <div
              class="MuiGrid-root MuiGrid-item css-z7x17r-MuiGrid-root"
            >
               
              <h4
                class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1vw6mcs-MuiTypography-root"
              >
                Would you rather?
              </h4>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item css-z7x17r-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-1j94i89-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for=":r1:"
                  id=":r1:-label"
                >
                  First
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r1:"
                    placeholder="3"
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-yjsfm1"
                    >
                      <span>
                        First
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item css-z7x17r-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-1j94i89-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for=":r2:"
                  id=":r2:-label"
                >
                  Second
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r2:"
                    placeholder="3"
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-yjsfm1"
                    >
                      <span>
                        Second
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item css-z7x17r-MuiGrid-root"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-u4gkmc-MuiButtonBase-root-MuiButton-root"
                data-testid="show-button"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiTypography-root MuiTypography-button css-1bvnr3d-MuiTypography-root"
                >
                  Save
                </span>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
